File: Python/cluster.py
import point
 
# Retorna o ponto Lider de um grupo
def group_leader(group):
    return group[0]

# Consiste em separar um conjunto de pontos em N grupos, sendo que
# os pontos pertencentes a um mesmo grupo possuem uma distancia euclidiana
# menor ou igual a um limite pre-determinado.
def clustering(points,limit):
    groups = [ [points[0]] ]

    for i in range(1,len(points)):
        leader = True

        for j in range(len(groups)):
            if (points[i].euclideanDistance(group_leader(groups[j])) <= limit):
                groups[j].append(points[i])
                leader = False
                break
        if leader:
            newGroup = [points[i]]
            groups.append(newGroup)
    return groups

# O centroide eh o ponto representativo de um grupo e eh calculado
# como o centro de massa do grupo.
def centroid (group):
    coords = []
    group_size = len(group)
    for i in range(len(group_leader(group).coordenates)):
        coords.append(0)
        for j in range(group_size):
            coords[i] += group[j].coordenates[i]
        coords[i] /= float(group_size)
    return point.Point(coords,0)

# Calcula a soma das distancias euclidianas quadradas(SSE) entre os pontos
# pertencentes a cada um dos grupos. Para tanto, eh utilizado o centroide de cada grupo.
def sse(groups):
    sum = 0.0

    for i in range(len(groups)):
        c = centroid(groups[i])
        for j in range(len(groups[i])):
            sum += (groups[i][j].euclideanDistance(c))**2
    return sum

File: Rust/cluster.rs
use crate::pts;

// Consiste em separar um conjunto de pontos em N grupos, sendo que
// os pontos pertencentes a um mesmo grupo possuem uma distancia euclidiana
// menor ou igual a um limite pre-determinado.
pub fn clustering (points: &Vec<pts::Point>,limit: f64) -> Vec<Vec<&pts::Point>> {
    let mut groups: Vec<Vec<&pts::Point>> = vec![vec![]];
	groups[0].push(&points[0]);

    for i in 1..points.len() {
		let mut leader = true;
		
		for j in 0..groups.len() {
			if pts::euclidean_distance(&points[i],&groups[j][0]) <= limit {
				groups[j].push(&points[i]);
				leader = false;
				break;
			}
		}
		if leader {
			groups.push(vec![ &points[i] ]);
		}
	}
    return groups;
}

// O centroide eh o ponto representativo de um grupo e eh calculado
// como o centro de massa do grupo.
fn centroid (points: &Vec<pts::Point>, group: &Vec<&pts::Point>) -> pts::Point {
    let mut coords: Vec<f64> = Vec::new(); 
    let len_coords = points[0].coordenates.len();
    let len_group = group.len();
   
    for i in 0..len_coords {
        coords.push(0.0);
        for j in 0..len_group {
            coords[i] += group[j].coordenates[i];
        }
        coords[i] /= len_group as f64;
    }
    return pts::Point{coordenates: coords, index: 0};
}

// Calcula a soma das distancias euclidianas quadradas(SSE) entre os pontos
// pertencentes a cada um dos grupos. Para tanto, eh utilizado o centroide de cada grupo.
pub fn sse (points: &Vec<pts::Point>, groups: &Vec<Vec<&pts::Point>>) -> f64{
    let mut sum = 0 as f64;

    for group in groups {
        let c: pts::Point = centroid(&points,group);
        for point in group {
            sum += pts::euclidean_distance(point,&c).powf(2.0);
        }
    }
    return sum;
}